modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/GroupedExternalResult.java

```diff
@@ -0,0 +1,236 @@
+        if (resQueue.isEmpty())
+            return null;
+
+        Chunk batch = resQueue.poll();
+
+        Object[] row = batch.currentRow();
+
+        if (batch.next())
+            resQueue.offer(batch);
+
+        return row;
+    }
+
+    /**
+     * @param data Groups to spill.
+     */
+    public void spillGroupsToDisk(TreeMap<ValueRow, Object[]> data) {
+        size += data.size();
+
+        ArrayList<Object[]> rows = new ArrayList<>(data.size());
+
+        for (Map.Entry<ValueRow, Object[]> e : data.entrySet()) {
+            ValueRow key = e.getKey();
+            Object[] aggs = e.getValue();
+
+            Object[] newRow = getObjectsArray(key, aggs);
+
+            rows.add(newRow);
+        }
+
+        Data buff = createDataBuffer(rowSize(rows));
+
+        long initFilePos = lastWrittenPos;
+
+        for (Object[] row : rows)
+            addRowToBuffer(row, buff);
+
+        setFilePosition(initFilePos);
+
+        long written = writeBufferToFile(buff);
+
+        lastWrittenPos = initFilePos + written;
+
+        chunks.add(new Chunk(initFilePos, lastWrittenPos));
+    }
+
+    /**
+     * Creates merged array of key and aggs.
+     * @param key Key.
+     * @param aggs Aggregates.
+     * @return Merged array.
+     */
+    @NotNull private Object[] getObjectsArray(ValueRow key, Object[] aggs) {
+        Object[] newRow = new Object[aggs.length + 1];
+
+        newRow[0] = key;
+
+        System.arraycopy(aggs, 0, newRow, 1, aggs.length);
+
+        return newRow;
+    }
+
+
+    /** Invoked after all rows have been spilled to disk.*/
+    public void reset() {
+        if (resQueue != null) {
+            resQueue.clear();
+
+            for (Chunk chunk : chunks)
+                chunk.reset();
+        }
+        else {
+            resQueue =  new PriorityQueue<>(new Comparator<Chunk>() {
+                @Override public int compare(Chunk o1, Chunk o2) {
+                    int c = cmp.compare((Value)o1.currentRow()[0], (Value)o2.currentRow()[0]);
+
+                    if (c != 0)
+                        return c;
+
+                    // Compare batches to ensure they emit rows in the arriving order.
+                    return Long.compare(o1.start, o2.start);
+                }
+            });
+        }
+
+        // Init chunks.
+        for (Chunk chunk : chunks) {
+            if (chunk.next())
+                resQueue.offer(chunk);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Object[] createEmptyArray(int colCnt) {
+        return new Object[colCnt];
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void onClose() {
+        super.onClose();
+    }
+
+    /**
+     * Extracts distinct row key from the row.
+     * @param row Row.
+     * @return Distinct key.
+     */
+    public ValueRow getRowKey(Object[] row) {
+        return (ValueRow)row[0];
+    }
+
+    /**
+     * Sorted rows chunk on the disk.
+     */
+    private class Chunk {
```

Comment:
Let's make this static as it doesn't use parent context.

Reply:
It uses `readRowFromFile()` from parent.